src/fhir.js

```diff
@@ -318,7 +321,17 @@ class Patient extends FHIRObject {
     const resourceType = classInfo.name.replace(/^FHIR\./, '');
     const records = this._bundle.entry
       .filter(e => {
-        return e.resource && e.resource.resourceType == resourceType;
+        if (e.resource && e.resource.resourceType == resourceType) {
+          if (this._shouldCheckProfile) {
```

Comment:
I think we really should only go into this if the `profile` is _not_ the same as the canonical URL for the resourceType.  E.g.,
```suggestion
          if (this._shouldCheckProfile && profile !== `http://hl7.org/fhir/StructureDefinition/${resourceType}`) {
```
... because, again, it's redundant.  If the profile is the base canonical URL, and the resource has the `resourceType` implied by the URL, then it's a match -- with or without `meta.profile`.

Reply:
Looking at the surrounding code, however... it really seems to me that this feature can't work as intended (for the _actual profile_ use case) with the one-arg implementation of `findRecords`.

In cql-execution, this gets invoked like this:
```
let records = ctx.findRecords(
  this.templateId != null ? this.templateId : this.datatype,
  retrieveDetails
);
```
Using a profile-aware model that then issues ELM using the FHIR model, I think the `templateId` would be the canonical URL of the _profile_.  For example, if we use the qicore model, then the retrieve's `templateId` would be `http://hl7.org/fhir/us/qicore/StructureDefinition/qicore-condition` -- and that's what would be passed in as the `profile`.

The problem is that in line 316 up above, it will pass that profile in to `this._modelInfo.findClass(profile)` -- but that won't be able to find it because it expects the `profile` to match a base FHIR resource (not a profile).

So... unless I'm missing something, this implementation only supports "profiles" that are actually just the base canonical for the resource; any non-base profile (like a QICore profile) will fail this implementation before it even gets to the code that looks at `meta.profile`.  But supporting only base resource canonicals is pointless because the base resource is already encoded in the `resourceType`.

It seems to me that the only way to properly support this is to update the function to consider the `retrieveDetails` argument -- so then it can use the retrieve's `dataType` to get the base resource type, and then the `templateId` to filter on `meta.profile` (as long as it is not the canonical URL for the base resource itself). This, of course, means you'd also need to use a version of cql-execution that supports passing in the `retrieveDetails`.

Reply:
Hah. Good point. Maybe I should include the changes we made that make use of `retrieveDetails`. I kinda forgot (my bad) that in order for this whole picture to work for us (whole picture being "calculating a measure authored `using QICore ...`), I believe we needed a combination of both retrieve detail usage and [optionally] meta.profile checking.

Reply:
So question then... should I also update this PR to include code for making use of that retrieve details object to get this to actually work with more profiles in practice? Of course, this could be a bit complex of a question because that means that this code will _only_ work with versions of cql-execution where retrieve details is included (AKA >=2.4.1).

That means the peer dependency of `cql-execution` in here would go from `>=1.3.0` to `>=2.4.1` to get this feature... is that even worth it? Does that constitute a breaking change for `cql-exec-fhir`, even though it's an opt-in thing that wouldn't actually break existing implementations, but technically trusted env wouldn't work at all unless you're using `cql-execution` 2.4.1 or newer... hmmm

Reply:
That was a lot of questions!  But I _think_ I can provide some answers (maybe):
1. I don't think it makes sense to merge this PR as-is.  So it probably makes sense to push up the other changes that use `retrieveDetails` to complete the functionality.
2. If those changes only affect functionality when `shouldCheckProfile` is enabled _AND_ the behavior otherwise continues to work as it did before, then I think this can be a minor version bump.  It's not a breaking change for existing implementations; it just introduces a feature that carries certain requirements along with it (if you want to use the new feature).
3. In this case, I think we'd need to keep the dependency at `>=1.3.0` to reflect minimum requirements, but... we need to very clearly document that the trusted environment _only_ works with a CQL 2.4.1+ server.  And if `shouldCheckProfile` is on and `findRecords` is called _without_ `retrieveDetails`, then it should emit an error saying that cql-execution 2.4.1 or above is required.

Does that make sense?